atef/report.py

```diff
@@ -0,0 +1,546 @@
+        bookmark_name = hashlib.sha1(mark_name).hexdigest()
+        header = Paragraph(text + f'<a name="{bookmark_name}"/>', style)
+        header._bookmark_name = bookmark_name
+        return header
+
+    def create_report(self) -> None:
+        """ Build the final report. """
+        raise NotImplementedError()
+
+
+class PassiveAtefReport(AtefReport):
+    """
+    Report for Passive Checkouts.  Assumes specific PreparedFile structure
+    """
+    __doc__ = __doc__ + AtefReport.__doc__
+
+    def create_report(self):
+        """
+        Top-level method for creating the final report.
+        Use the stored config and calls various helpers
+        """
+        # Build story as a list of Flowable objects
+        story = []
+        # Cover page
+        self.build_cover_page(story)
+        # Results summary page
+        self.build_summary(story)
+        # Individual config / comparison pages
+        for c, _ in self.walk_config_file(self.config.root):
+            self.build_config_page(story, c)
+
+        # must build several times to place items then gather info for ToC
+        self.multiBuild(story)
+
+    def build_summary(self, story: List[Flowable]):
+        """
+        Build summary table for checkout
+
+        Parameters
+        ----------
+        story : List[Flowable]
+            a list of components used to render the report.  New items
+            are appended to this directly
+        """
+        story.append(self.build_linked_header('Checkout Summary', h1))
+        story.append(platypus.Spacer(width=0, height=.5*cm))
+        # table with results
+        lines = list(self.walk_config_file(self.config.root))
+        table_data = [['Step Name', 'Result']]
+        style = [('VALIGN', (0, 0), (-1, -1), 'TOP'),
+                 ('ALIGN', (0, 0), (1, 0), 'CENTER'),
+                 ('BOX', (0, 0), (-1, -1), 1, colors.black),
+                 ('BOX', (0, 0), (0, -1), 1, colors.black)]
+
+        for i in range(len(lines)):
+            # content
+            item, level = lines[i]
+            prefix = '    ' * level
+            if isinstance(item, PreparedConfiguration):
+                name = item.config.name
+            elif isinstance(item, PreparedComparison):
+                name = item.comparison.name
+            name = name or type(item).__name__
+            table_data.append(
+                [
+                    prefix + f'{name}',
+                    self.get_result_text(item.result)
+                ]
+            )
+
+            # style
+            if isinstance(item, PreparedConfiguration):
+                style.append(['LINEABOVE', (0, i+1), (-1, i+1), 1, colors.black])
+            else:
+                style.append(['LINEABOVE', (0, i+1), (-1, i+1), 1, colors.lightgrey])
+
+        table = platypus.Table(
+            table_data, style=style
+        )
+
+        story.append(table)
+        story.append(platypus.PageBreak())
+
+    def walk_config_file(
```

Comment:
Hmm, next refactor opportunity: this function doesn't seem to belong here. It could be out in a utility somewhere as it doesn't rely on `self`